### PR TITLE
FM2-411: Added support for instance level $everything operation on Patients in R4

### PR DESCRIPTION
**Pull request by Medhavi-16**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-411

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by Medhavi-16** [comment]
@ibacher @varung-31 PTAL.

**Review by varung-31 — CHANGES_REQUESTED** [review]
Thanks for your work on this @Medhavi-16 
I have added some comments for refactoring, but the core logic looks good!

File: api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQuery.java [hunk 1]
```diff
@@ -37,10 +39,20 @@
...
 	 */
 	public IBundleProvider getQueryResults(SearchParameterMap theParams, O dao, V translator, W searchQueryInclude) {
+		if (!theParams.getParameters(FhirConstants.PATIENT_EVERYTHING_SEARCH_HANDLER).isEmpty()) {
+			SimpleBundleProvider result = new SimpleBundleProvider(
+			        new SearchQueryBundleProvider<>(theParams, dao, translator, globalPropertyService, searchQueryInclude)
+			                .getAllResources());
+			
+			result.setPreferredPageSize(result.size());
+			return result;
+		}
```
[the remark this answers is not in the document]
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQuery.java:55]: @varung-31 I was thinking along the lines that when in future we implement $everything operation on other resources(eg: Encounter), we'd have to prevent pagination there also, so maybe adding it here would make it useable for $everything on other resources as well.

[the remark this answers is not in the document]
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQuery.java:55]: This would make more sense if we generalise the FhirConstants.PATIENT_EVERYTHING_SEARCH_HANDLER to just EVERYTHING_SEARCH_HANDLER and change its value as well

[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQuery.java:55]: I'm not sure why we're trying to wrap things in a `SimpleBundleProvider`... that seems unnecessary when we could just add this logic to the `SearchQueryBundleProvider`'s `preferredPageSize()` implementation? I tend to agree that having a generic `EVERYTHING_SEARCH_HANDLER` seems like the way forward. Alternatively, I suppose this could just be done in the `BaseFhirService`....

[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQuery.java:55]: Yup, we can incorporate the change in the preferred page size method as well by checking for a generic `EVERYTHING_SEARCH_HANDLER `

[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/SearchQuery.java:55]: I see. Yes, you're absolutely right. Thanks for the explanation!

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java [hunk 2]
```diff
@@ -143,4 +146,18 @@ public IBundleProvider searchPatients(@OptionalParam(name = Patient.SP_NAME) Str
...
 		    city, state, postalCode, country, id, lastUpdated, sort, revIncludes);
 	}
+
+	/**
+	 * The $everything operation fetches all the information related the specified patient
+	 * @param patientId The id of the patient
+	 * @return a bundle of resources which reference to or are referenced from the patient
+	 */
+	@Operation(name = "everything", idempotent = true, type = Patient.class, bundleType = BundleTypeEnum.SEARCHSET)
+	public IBundleProvider getPatientEverything(@IdParam IdType patientId) {
```
[the remark this answers is not in the document]
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: @varung-31 @ibacher PTAL.

**Comment by ibacher** [comment]
@Medhavi-16 Looks like you accidentally left references to `PATIENT_EVERYTHING_SEARCH_HANDLER` in the tests.

**Comment by Medhavi-16** [comment]
> @Medhavi-16 Looks like you accidentally left references to `PATIENT_EVERYTHING_SEARCH_HANDLER` in the tests.

Yeah, sorry, I missed to commit the files.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java [hunk 3]
```diff
@@ -106,4 +114,26 @@ public IBundleProvider searchForPatients(StringAndListParam name, StringAndListP
...
+		
+		HashSet<Include> revIncludes = new HashSet<>();
+		
+		revIncludes.add(new Include("Observation:" + FhirConstants.INCLUDE_PATIENT_PARAM));
+		revIncludes.add(new Include("AllergyIntolerance:" + FhirConstants.INCLUDE_PATIENT_PARAM));
+		revIncludes.add(new Include("DiagnosticReport:" + FhirConstants.INCLUDE_PATIENT_PARAM));
+		revIncludes.add(new Include("Encounter:" + FhirConstants.INCLUDE_PATIENT_PARAM));
+		revIncludes.add(new Include("MedicationRequest:" + FhirConstants.INCLUDE_PATIENT_PARAM));
+		revIncludes.add(new Include("ServiceRequest:" + FhirConstants.INCLUDE_PATIENT_PARAM));
+		revIncludes.add(new Include("ProcedureRequest:" + FhirConstants.INCLUDE_PATIENT_PARAM));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:133]: Can we use the `FhirConstants` constants for these too? E.g.,

```java
revIncludes.add(new Include(FhirConstants.OBSERVATION + ":" + FhirConstants.INCLUDE_PATIENT_PARAM));
```

Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:133]: Yes, I'll change it to that.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java [hunk 4]
```diff
@@ -106,4 +107,26 @@ public IBundleProvider searchForPatients(StringAndListParam name, StringAndListP
 		
 		return searchQuery.getQueryResults(theParams, dao, translator, searchQueryInclude);
 	}
+	
+	@Override
+	@Transactional(readOnly = true)
+	public IBundleProvider getPatientEverything(TokenAndListParam patientId) {
+		SearchParameterMap theParams = new SearchParameterMap()
+		        .addParameter(FhirConstants.EVERYTHING_SEARCH_HANDLER, new StringParam())
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:115]: Since the value here doesn't matter and `SearchParameterMap` takes any `Serializable` object as a valid value, this is a more sensible implementation:

```suggestion
		        .addParameter(FhirConstants.EVERYTHING_SEARCH_HANDLER, "")
```

(The idea for requesting this change is that `new StringParam()` creates a object whereas `""` already exists in the string constant pool, so we save a tiny fraction of memory. It also, I think, makes it a little clearer that the value here isn't meaningful).

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: @Medhavi-16 I'd do the `null` check something like this:

```java
if (patientId == null) {
    return null;
}
```

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java [hunk 5]
```diff
@@ -106,4 +107,26 @@ public IBundleProvider searchForPatients(StringAndListParam name, StringAndListP
 		
 		return searchQuery.getQueryResults(theParams, dao, translator, searchQueryInclude);
 	}
+	
+	@Override
+	@Transactional(readOnly = true)
+	public IBundleProvider getPatientEverything(TokenAndListParam patientId) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:113]: Why do we take a `TokenAndListParam` here? Aren't we always expected to have a single patient? (I get that we need to pass a `TokenAndListParam` as the search parameter; I just think we should construct the one-value `AndListParam` here instead of in the `ResourceProvider`... that way, the interface for the service method is clear that it *only* expects a single value).

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: PS With my other suggestions, this can then be simplified to:

```java
TokenParam patientReference = new TokenParam(patientId.getIdPart());
```

[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: @ibacher I think for the null check, `(id == null || id.getIdPart() == null)` should be better.

[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: Even better might be `(id == null || id.getIdPart() == null || id.getIdPart().isEmpty())`

[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: I have a small question here: How do we actually supply null or empty values in a path variable?
Because trying to hit the `https://hapi.fhir.org/baseR4/Patient//$everything` URL redirects me to the type-level $everything operation instead.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java (hunk 5 above)
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:113]: @ibacher Should we create and pass just a `TokenParam` from resource provider to service?

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:113]: Yes

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProvider.java:156]: So I don't think there's an explicit way to craft a URL that does that or if there were, it would be an error somewhere in HAPI's code base. But if such an error is accidentally introduced at some point, I'd prefer we do something sensible here rather than send something far down the chain and, e.g., run a query to select all of the patients and everything associated with them.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java (hunk 4 above)
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java]: @ibacher Should we create a PR to correct this for `$lastn` operations because I had done a similar thing there?

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java]: Yeah, I suppose that's a good idea.

**Comment by Medhavi-16** [comment]
@ibacher @varung-31 PTAL.

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProviderTest.java [hunk 6]
```diff
@@ -528,7 +528,27 @@ public void searchForPatients_shouldNotAddResourcesForEmptyRevInclude() {
 		assertThat(((Patient) resultList.iterator().next()).getId(), equalTo(PATIENT_UUID));
 	}
 	
+	@Test
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/PatientFhirResourceProviderTest.java:531]: We can add a test for null patient id as an input.

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImplTest.java [hunk 7]
```diff
@@ -794,6 +794,43 @@ public void searchForPatients_shouldNotAddRelatedResourcesToResultListForEmptyRe
 		assertThat(resultList.size(), greaterThanOrEqualTo(1));
 	}
 	
+	@Test
+	public void getPatientEverything_shouldReturnAllInformationAboutSpecifiedPatient() {
+		TokenParam patientId = new TokenParam().setValue(PATIENT_UUID);
+		
+		SearchParameterMap theParams = new SearchParameterMap()
+		        .addParameter(FhirConstants.EVERYTHING_SEARCH_HANDLER, new StringParam())
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImplTest.java:802]:
```suggestion
		        .addParameter(FhirConstants.EVERYTHING_SEARCH_HANDLER, "")
```

File: api/src/test/java/org/openmrs/module/fhir2/api/search/PatientSearchQueryTest.java [hunk 8]
```diff
@@ -1154,4 +1158,34 @@ public void shouldReturnCollectionOfPatientsSortedByCountry() {
 		}
 	}
 	
+	@Test
+	public void searchForPatient_shouldReturnPatientEverything() {
+		TokenAndListParam patientId = new TokenAndListParam().addAnd(new TokenParam().setValue(PATIENT_OTHER2_UUID));
+		
+		SearchParameterMap theParams = new SearchParameterMap()
+		        .addParameter(FhirConstants.EVERYTHING_SEARCH_HANDLER, new StringParam())
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/api/search/PatientSearchQueryTest.java:1166]:
```suggestion
		        .addParameter(FhirConstants.EVERYTHING_SEARCH_HANDLER, "")
```

**Review by varung-31 — CHANGES_REQUESTED** [review]

**Review by ibacher — APPROVED** [review]
LGTM! Thanks @Medhavi-16!

**Comment by ibacher** [comment]
Whoops... forgot to check the tests... @Medhavi-16 can you fix the test issue? (Basically, you just need to remove the `when()` calls from the new tests; see [the log](https://github.com/openmrs/openmrs-module-fhir2/pull/360/checks?check_run_id=3127933811#step:5:405) for too many details).

**Comment by Medhavi-16** [comment]
> Whoops... forgot to check the tests... @Medhavi-16 can you fix the test issue? (Basically, you just need to remove the `when()` calls from the new tests; see [the log](https://github.com/openmrs/openmrs-module-fhir2/pull/360/checks?check_run_id=3127933811#step:5:405) for too many details).

Done @ibacher
